README.md
# Dotnetframework cookbook

Installs and configures the .NET framework 4, 4.5, 4.5.1, 4.5.2, or 4.6 runtime

# Requirements

Tested on Windows Server 2008 R2 and Windows Server 2012R2. The selected .NET
runtime should work on versions of Windows supported by the associated .NET
installer.

* Windows 2008
* Windows 2008 R2
* Windows 2012
* Windows 2012 R2

# Usage

Include the default recipe in your run list. The default recipe will install
the specified .NET framework version.

# Attributes

## default

* `node['dotnetframework']['version']` - defaults to '4.5.2' Acceptable values:
'4.0', '4.5', '4.5.1', '4.5.2', '4.6', '4.6.1', '4.6.2'.

# Recipes

## default

Installs the .NET Framework.

## regiis

This recipe register .NET with IIS so that IIS can host .NET application
associated with the specified intalled .NET version. This recipe currently
does not support Windows 2012 or higher. For Windows 2012 or newer its
recommended that you use the IIS cookbook to register the .NET version.

## mini-tests

You can include the mini-tests in your runlist to verify .NET was successfully
installed, however .NET will not work until you reboot.

.NET 4.6 minitests will fail until you reboot, so its best to run Chef with
only dotnetframework in your runlist, reboot, then include dotnetframework
again with the minitest-handler.

# TODO

- Install .NET using the windows_feature resource if the current OS supports it.
- Support older versions of .NET < 4.0.
- Support installation over WinRM (i.e. native scheduled task support).
- Abstract .NET installation to a Chef resource.

# Author

Author:: Shawn Neal (sneal@sneal.net)


CHANGELOG.md
# CHANGELOG for dotnetframework

This file is used to list changes made in each version of dotnetframework.

## 2.0.0
- ChefDK 2.5.3 rubocop and foodcritic updates
- Remove `node.set` for Chef 13 compatiblity

## 1.2.0
- .NET 4.6.1 support
- Installation is skipped if a newer .NET version is already installed.
- Mini-tests do not fail if a newer .NET version is already installed.

## 1.1.0
- Fixed .NET framework idempotence to avoid constant reboots
- Added dotnetframework_version LWRP

## 1.0.0:
- Initial version released on Supermarket.

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

ChefSpec::Coverage.start!

RSpec.configure do |config|
  config.run_all_when_everything_filtered = true
  config.filter_run :focus
  config.order = 'random'
  config.platform = 'windows'
  config.version = '2012R2'

  config.before(:each) do
    # Set some common global ENV vars used by Windows cookbooks
    ENV['WINDIR'] = 'C:\Windows'
    stub_const('File::ALT_SEPARATOR', '\\')
  end
end


spec/default_spec.rb
describe 'dotnetframework::default' do
  describe '.NET is not installed' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new.converge(described_recipe)
    end
    it 'should install .NET 4.5.2' do
      expect(chef_run).to install_dotnetframework_version('4.5.51650')
    end
    it 'should notify reboot resource' do
      win_pkg = chef_run.dotnetframework_version('4.5.51650')
      expect(win_pkg).to notify(
        'reboot[dotnetframework_install]'
      ).to(:request_reboot).immediately
    end
  end
end


files/default/tests/minitest/default_test.rb
# encoding: UTF-8

require 'minitest/spec'
require 'win32/registry'

# Tests to ensure .NET 4 is installed
class TestDotNet4Install < MiniTest::Chef::TestCase
  def test_framework_dir_exists
    # all .NET versions are installed to the same v4.0.30319 folder
    dir = File.join(ENV['WINDIR'], 'Microsoft.Net\\Framework64\\v4.0.30319')
    assert Dir.exist?(dir)
  end

  # Determining installed .NET versions
  # http://support.microsoft.com/kb/318785
  # https://msdn.microsoft.com/en-us/library/hh925568(v=vs.110).aspx
  def test_framework_version
    expected_version = expected_dotnet_version
    reg_version = installed_dotnet_version
    flunk('Could not find a .NET version in the registry') unless reg_version
    assert(
      Gem::Version.new(reg_version) >= Gem::Version.new(expected_version),
      "Expected .NET version #{expected_version} or higher, " \
      "but only found #{reg_version}"
    )
  end

  def expected_dotnet_version
    major_version = node['dotnetframework']['version']
    node['dotnetframework'][major_version]['version']
  end

  def installed_dotnet_version
    path = 'SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full'
    Win32::Registry::HKEY_LOCAL_MACHINE.open(path) do |reg|
      return reg['Version']
    end
  rescue ::Win32::Registry::Error
    return nil
  end
end


rakefile.rb
# rubocop:disable Metrics/BlockLength
task default: %i[version rubocop foodcritic spec kitchen cleanup]
task nokitchen: %i[version rubocop foodcritic spec]

desc 'Set cookbook version.'
task :version do
  version = ENV['BUILD_NUMBER'] ? "2.0.#{ENV['BUILD_NUMBER']}" : '2.0.0'
  IO.write('version.txt', version)
end

desc 'Foodcritic test.'
task :foodcritic do
  load_gems('colorize')
  puts 'Running Foodcritic tests...'.green
  sh 'chef exec foodcritic . -G -f any'
end

desc 'Rubocop test.'
task :rubocop do
  load_gems('colorize')
  puts 'Running Rubocop tests...'.green
  sh 'chef exec rubocop'
end

desc 'ChefSpec test.'
task :spec do
  load_gems('colorize minitest-chef-handler')
  puts 'Running ChefSpec tests...'.green
  cmd = 'chef exec rspec --color -f documentation ' \
   '-tunit --pattern "spec/**/*_spec.rb"'
  sh cmd.to_s
end

desc 'Test Kitchen.'
task :kitchen, [:type] => %i[check_env_vars assume_role] do |_t, args|
  task_thread = []
  kitchen_complete = false
  start_time = Time.now
  %w[kitchen assume_role].each do |task|
    task_thread << Thread.new do
      case task
      when 'kitchen'
        load_gems('colorize')
        puts 'Running Test Kitchen...'.green
        type = args[:type] || 'test'
        begin
          sh "chef exec kitchen #{type} -c"
        rescue StandardError => e
          Rake::Task[:cleanup].invoke if type == 'test'
          raise e.inspect
        end
        kitchen_complete = true
      when 'assume_role'
        while kitchen_complete == false
          current_time = Time.now
          elapsed_time = (current_time.to_f - start_time.to_f).to_i
          if elapsed_time > 3000
            Rake::Task[:assume_role].reenable
            Rake::Task[:assume_role].invoke
            start_time = Time.now
          end
          sleep 10
        end
      end
    end
  end
  task_thread.each(&:join)
end

desc 'Cleanup after yourself!'
task :cleanup, [:cookbook_name] => [:assume_role] do |_t, args|
  load_gems('colorize')
  puts "Clean up AWS in case test kitchen doesn't. . .".green
  cookbook_name = args.cookbook_name || File.basename(Dir.getwd)
  instance_cmd = 'aws ec2 describe-instances --filters ' \
   "\"Name=tag:Name,Values=#{cookbook_name}\" " \
   '"Name=instance-state-name,Values=running"'

  instance = `#{instance_cmd}`
  unless JSON.parse(instance)['Reservations'].empty?
    id = JSON.parse(instance)['Reservations'][0]['Instances'][0]['InstanceId']
    sh "aws ec2 terminate-instances --instance-ids #{id}"
  end
  puts "Cookbook #{cookbook_name} is cleaned up.".green
end

desc 'AWS Assume Role.'
task assume_role: :check_env_vars do
  load_gems('json colorize')

  puts "Assuming role for #{ENV['EC2_ROLE_ARN']}".green

  assume_role_cmd = 'aws sts assume-role ' \
    "--role-arn=#{ENV['EC2_ROLE_ARN']} " \
    '--role-session-name=temp_session'
  data = JSON.parse(`#{assume_role_cmd}`)

  access_key = data['Credentials']['AccessKeyId']
  secret_key = data['Credentials']['SecretAccessKey']
  token = data['Credentials']['SessionToken']

  system "aws configure set aws_access_key_id #{access_key}"
  system "aws configure set aws_secret_access_key #{secret_key}"
  system "aws configure set aws_session_token #{token}"
  system "aws configure set region #{ENV['EC2_REGION']}"

  ENV['AWS_ACCESS_KEY_ID'] = nil
  ENV['AWS_SECRET_ACCESS_KEY'] = nil

  puts "Role '#{ENV['EC2_ROLE_ARN']}' has been assumed.".green
end

desc 'Verify environment variables are set.'
task :check_env_vars do
  load_gems('colorize')
  @vars = []
  @file = File.new('docker-compose.yml', 'r')
  while (line = @file.gets)
    @vars << line.scan(/^\s+\b[A-Z0-9_]+:\s+\${(\w+)/).join('')
  end
  req_env_vars = @vars.reject(&:empty?)
  msg = ''
  req_env_vars.uniq.each do |env|
    next if env.include? '_DIR'
    msg += "\n#{env}" if ENV[env].to_s.empty?
  end
  raise 'Missing Environment Variables. . .'.red + msg.red unless msg.empty?
end

private

def load_gems(gems)
  gems.split(' ').each do |g|
    begin
      gem g
    rescue Gem::LoadError
      system "gem install #{g} --no-document"
      Gem.clear_paths
    end
    require g
  end
end


recipes/regiis.rb
#
# Author:: Daptiv Engineering (<cpc_sea_teamengineering@changepoint.com>)
# Cookbook Name:: dotnetframework
# Recipe:: regiis
#
# Copyright:: Copyright (c) 2018 Changepoint
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# rubocop:disable Method/LineLength
# C:\Windows\System32\inetsrv>appcmd list config /section:isapiCgiRestriction
# <system.webServer>
#   <security>
#     <isapiCgiRestriction>
#       <add path="%windir%\Microsoft.NET\Framework64\v2.0.50727\aspnet_isapi.dll" \
#         allowed="true" groupId="ASP.NET v2.0.50727" />
#       <add path="%windir%\Microsoft.NET\Framework\v2.0.50727\aspnet_isapi.dll" \
#         allowed="true" groupId="ASP.NET v2.0.50727" />
#     </isapiCgiRestriction>
#   </security>
# </system.webServer>

dotnet4dir = File.join(ENV['WINDIR'], 'Microsoft.Net\\Framework64\\v4.0.30319')
aspnet_regiis_cmd = File.join(dotnet4dir, 'aspnet_regiis.exe')
appcmd = File.join(ENV['WINDIR'], 'system32\\inetsrv\\appcmd.exe')

execute 'aspnet_regiis' do
  command "#{aspnet_regiis_cmd} -i"
  action :run
  not_if do
    cmd = Mixlib::ShellOut.new("#{appcmd} list config /section:isapiCgiRestriction")
    aspnet_isapis = cmd.run_command
    Chef::Log.debug(aspnet_isapis.stdout)
    aspnet_isapis.stdout.include?('Framework64\\v4.0.30319\\aspnet_isapi.dll')
  end
end


resources/version.rb
# encoding: UTF-8

# Author:: Daptiv Engineering (<cpc_sea_teamengineering@changepoint.com>)
# Cookbook Name:: dotnetframework
# Resource:: edition
#
# Copyright 2018 Changepoint
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

actions :install
default_action :install

# The full .NET version: 4.0.30319, 4.5.51641 etc
attribute :version, kind_of: String, name_attribute: true

# The offline installer package URL
attribute :source, kind_of: String, required: true

# The installed MSI package name
attribute :package_name, kind_of: String, required: true

# The source ISO SHA256 checksum
attribute :checksum, kind_of: String


libraries/matchers.rb
# encoding: UTF-8

# Author:: Daptiv Engineering (<cpc_sea_teamengineering@changepoint.com>)
# Cookbook Name:: dotnetframework
#
# Copyright 2018 Changepoint
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if defined?(ChefSpec)
  chefspec_version = Gem.loaded_specs['chefspec'].version
  define_method = \
    if chefspec_version < Gem::Version.new('4.1.0')
      ChefSpec::Runner.method(:define_runner_method)
    else
      ChefSpec.method(:define_matcher)
    end

  define_method.call :dotnetframework_version

  def install_dotnetframework_version(resource_name)
    ChefSpec::Matchers::ResourceMatcher.new(
      :dotnetframework_version,
      :install,
      resource_name
    )
  end
end


attributes/default.rb
# encoding: UTF-8

# Author:: Daptiv Engineering (<cpc_sea_teamengineering@changepoint.com>)
# Cookbook Name:: dotnetframework
# Attribute:: default
#
# Copyright:: Copyright (c) 2018 Changepoint
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# rubocop:disable VariableNumber
# Allows the cookbook the compile without error on Linux
if RUBY_PLATFORM =~ /mswin|mingw32|windows/
  require 'chef/win32/version'
  windows_version = Chef::ReservedNames::Win32::Version.new
  is_2012r2_or_8_1 = \
    windows_version.windows_server_2012_r2? || windows_version.windows_8_1?
else
  is_2012r2_or_8_1 = true
end

# Override to install different .NET versions
# 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1
default['dotnetframework']['version'] = '4.5.2'

# .NET 4.0
default['dotnetframework']['4.0']['package_name'] =
  'Microsoft .NET Framework 4 Extended'
default['dotnetframework']['4.0']['version'] = '4.0.30319'
default['dotnetframework']['4.0']['checksum'] =
  '65e064258f2e418816b304f646ff9e87af101e4c9552ab064bb74d281c38659f'
default['dotnetframework']['4.0']['url'] =
  'http://download.microsoft.com/download/9/5/A/' \
  '95A9616B-7A37-4AF6-BC36-D6EA96C8DAAE/' \
  'dotNetFx40_Full_x86_x64.exe'

# .NET 4.5
default['dotnetframework']['4.5']['package_name'] =
  'Microsoft .NET Framework 4.5'
default['dotnetframework']['4.5']['version'] = '4.5.50709'
default['dotnetframework']['4.5']['checksum'] =
  'a04d40e217b97326d46117d961ec4eda455e087b90637cb33dd6cc4a2c228d83'
default['dotnetframework']['4.5']['url'] =
  'http://download.microsoft.com/download/b/a/4/' \
  'ba4a7e71-2906-4b2d-a0e1-80cf16844f5f/' \
  'dotnetfx45_full_x86_x64.exe'

# .NET 4.5.1
default['dotnetframework']['4.5.1']['package_name'] =
  'Microsoft .NET Framework 4.5.1'
if is_2012r2_or_8_1
  default['dotnetframework']['4.5.1']['version'] = '4.5.51641'
else
  default['dotnetframework']['4.5.1']['version'] = '4.5.50938'
end
default['dotnetframework']['4.5.1']['checksum'] =
  '5ded8628ce233a5afa8e0efc19ad34690f05e9bb492f2ed0413508546af890fe'
default['dotnetframework']['4.5.1']['url'] =
  'http://download.microsoft.com/download/1/6/7/' \
  '167F0D79-9317-48AE-AEDB-17120579F8E2/' \
  'NDP451-KB2858728-x86-x64-AllOS-ENU.exe'

# .NET 4.5.2
default['dotnetframework']['4.5.2']['package_name'] =
  'Microsoft .NET Framework 4.5.2'
if is_2012r2_or_8_1
  default['dotnetframework']['4.5.2']['version'] = '4.5.51650'
else
  default['dotnetframework']['4.5.2']['version'] = '4.5.51209'
end
default['dotnetframework']['4.5.2']['checksum'] =
  '6c2c589132e830a185c5f40f82042bee3022e721a216680bd9b3995ba86f3781'
default['dotnetframework']['4.5.2']['url'] =
  'http://download.microsoft.com/download/E/2/1/' \
  'E21644B5-2DF2-47C2-91BD-63C560427900/' \
  'NDP452-KB2901907-x86-x64-AllOS-ENU.exe'

# .NET 4.6
default['dotnetframework']['4.6']['package_name'] =
  'Microsoft .NET Framework 4.6'
default['dotnetframework']['4.6']['version'] = '4.6.00081'
default['dotnetframework']['4.6']['checksum'] =
  'b21d33135e67e3486b154b11f7961d8e1cfd7a603267fb60febb4a6feab5cf87'
default['dotnetframework']['4.6']['url'] =
  'http://download.microsoft.com/download/C/3/A/' \
  'C3A5200B-D33C-47E9-9D70-2F7C65DAAD94/' \
  'NDP46-KB3045557-x86-x64-AllOS-ENU.exe'

# .NET 4.6.1
default['dotnetframework']['4.6.1']['package_name'] =
  'Microsoft .NET Framework 4.6.1'
default['dotnetframework']['4.6.1']['version'] = '4.6.01038'
default['dotnetframework']['4.6.1']['checksum'] =
  'beaa901e07347d056efe04e8961d5546c7518fab9246892178505a7ba631c301'
default['dotnetframework']['4.6.1']['url'] =
  'http://download.microsoft.com/download/E/4/1/' \
  'E4173890-A24A-4936-9FC9-AF930FE3FA40/' \
  'NDP461-KB3102436-x86-x64-AllOS-ENU.exe'

# .NET 4.6.2
default['dotnetframework']['4.6.2']['package_name'] =
  'Microsoft .NET Framework 4.6.2'
default['dotnetframework']['4.6.2']['version'] = '4.6.01590'
default['dotnetframework']['4.6.2']['checksum'] =
  '28886593e3b32f018241a4c0b745e564526dbb3295cb2635944e3a393f4278d4'
default['dotnetframework']['4.6.2']['url'] =
  'https://download.microsoft.com/download/F/9/4/' \
  'F942F07D-F26F-4F30-B4E3-EBD54FABA377/' \
  'NDP462-KB3151800-x86-x64-AllOS-ENU.exe'
